src_c/camera_windows.c

```diff
@@ -0,0 +1,862 @@
+
+    hr = MFCreateAttributes(&pAttributes, 1);
+    HANDLEHR(hr);
+
+    hr = pAttributes->lpVtbl->SetGUID(pAttributes,
+                                      &MF_DEVSOURCE_ATTRIBUTE_SOURCE_TYPE,
+                                      &DEVSOURCE_VIDCAP_GUID);
+    HANDLEHR(hr);
+
+    hr = MFEnumDeviceSources(pAttributes, &ppDevices, &count);
+    HANDLEHR(hr);
+
+    for (int i=0; i<(int)count; i++) {
+        if (!_get_name_from_activate(ppDevices[i], &_device_name)) {
+            hr = HR_UPSTREAM_FAILURE;
+            goto cleanup;
+        }
+        if (!wcscmp(_device_name, device_name)) {
+            free(_device_name);
+            for (int j=0; j<(int)count; j++) {
+                if (i != j) {
+                    RELEASE(ppDevices[j]);
+                }
+            }
+            CoTaskMemFree(ppDevices);
+            return ppDevices[i];
+        }
+        free(_device_name);
+    }
+
+    goto cleanup;
+
+    cleanup:
+        RELEASE(pAttributes);
+        for (int i=0; i<(int)count; i++) {
+            RELEASE(ppDevices[i]);
+        }
+        CoTaskMemFree(ppDevices);
+        return NULL;
+}
+
+int
+_select_source_type(pgCameraObject* self, IMFMediaType **mp)
+{
+    HRESULT hr;
+    IMFMediaType* media_type = NULL;
+    int type_count = 0;
+    UINT32 t_width, t_height;
+    IMFMediaType** native_types = NULL;
+    int *diagonal_distances = NULL;
+    UINT64 size;
+    GUID subtype;
+    int difference = 100000000;
+    int current_difference;
+    int index = -1;
+    int requested_diagonal = DISTANCE(self->width, self->height);
+    UINT64 fps_max_ratio;
+
+    /* Find out how many native media types there are
+     * (different resolution modes, mostly) */
+    while(1) {
+        hr = self->reader->lpVtbl->GetNativeMediaType(self->reader,
+                                                      FIRST_VIDEO,
+                                                      type_count,
+                                                      &media_type);
+        if (hr) {
+            break;
+        }
+        type_count++;
+        RELEASE(media_type);
+    }
+
+    if (!type_count) {
+        PyErr_SetString(pgExc_SDLError,
+                        "Could not find any video types on this camera");
+        return 0;
+    }
+
+    native_types = malloc(sizeof(IMFMediaType*) * type_count);
+    diagonal_distances = malloc(sizeof(int) * type_count);
+    if (!native_types || !diagonal_distances) {
+        hr = E_OUTOFMEMORY;
+        HANDLEHR(hr);
+    }
+
+    /* This is interesting
+     * https://social.msdn.microsoft.com/Forums/windowsdesktop/en-US/9d6a8704-764f-46df-a41c-8e9d84f7f0f3/mjpg-encoded-media-type-is-not-available-for-usbuvc-webcameras-after-windows-10-version-1607-os
+     * Compressed video modes listed below are "backwards compatibility modes"
+     * And will always have an uncompressed counterpart (NV12 / YUY2)
+     * At least in Windows 10 since 2016 */
+
+    for (int i=0; i < type_count; i++) {
```

Comment:
A readability improvement could be to put these for loops into separate functions.